### ISSUE TITLE
Handle the sjdbOverhang parameter to accommodate the different read length per batch

### ISSUE THREAD

**Issue by hsun3163** [cumc/xqtl-pipeline#318]
verify 3.2.1 for following link:
https://physiology.med.cornell.edu/faculty/skrabanek/lab/angsd/lecture_notes/STARmanual.pdf

**Comment by hsun3163** [comment]
There are several obstacles to implementing the customized batch size.
1. The star we ran is within an embedded script from gTEX, which did not provide the option to specify the "sjdbOverhang" option.
2. As suspected, the `--sjdbOverhang`  specified in the second step will not overwrite the first step. Instead, an error will be thrown.
2.1 In the index file, the `--sjdbOverhang` is saved in the genomeParameter.txt file, as shown below. 

```
hs3163@node67:/mnt/vast/hpc/csg/xqtl_workflow_testing/rna_quant_hans/data/STAR_Index$ cat genomeParameters.txt
### STAR   --runMode genomeGenerate      --runThreadN 8   --genomeDir STAR_Index   --genomeFastaFiles /mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/data/GRCh38_full_analysis_set_plus_decoy_hla.fa      --sjdbGTFfile /mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/data/Homo_sapiens.GRCh38.103.chr.reformated.collapsed.gtf   --sjdbOverhang 150
### GstrandBit 33
versionGenome   2.7.4a
genomeType      Full
genomeFastaFiles        /mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/data/GRCh38_full_analysis_set_plus_decoy_hla.fa
genomeSAindexNbases     14
genomeChrBinNbits       18
genomeSAsparseD 1
genomeTransformType     None
genomeTransformVCF      -
sjdbOverhang    150
sjdbFileChrStartEnd     -
sjdbGTFfile     /mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/data/Homo_sapiens.GRCh38.103.chr.reformated.collapsed.gtf
sjdbGTFchrPrefix        -
sjdbGTFfeatureExon      exon
sjdbGTFtagExonParentTranscript  transcript_id
sjdbGTFtagExonParentGene        gene_id
sjdbInsertSave  Basic
genomeFileSizes 4125672917 26569342121
```
Changing this file is effectively overwriting the associated parameter.

However, I fail to see a way to implement it in our pipeline. As it is unwise to change the file in the index folder on the fly.

**Comment by gaow** [comment]
> Changing this file is effectively overwriting the associated parameter.

sounds like you have changed it manually and it works for the next step STAR command with `--sjdbOverhang` ? Even if it works, it does not sound right to use such a hack.

> . Instead, an error will be thrown.

let's always document the exact error message in tickets.

I think the key is to NOT set anything in the first step when reference index is generated. From section 3.2.1 : 

> Since 2.4.1a, the annotations can be included on the fly at the mapping step, without including
them at the genome generation step.

"without including" in the index step seems to be the key to not cause a conflict later -- could you try that?

**Comment by hsun3163** [comment]
> > Changing this file is effectively overwriting the associated parameter.
> 
> sounds like you have changed it manually and it works for the next step STAR command with `--sjdbOverhang` ? Even if it works, it does not sound right to use such a hack.

1. Yes, as long as there is no conflict between the set `--sjdbOverhang` parameter and what was in that file, there wont be error. Then again, we should not employ these hacks.

> > . Instead, an error will be thrown.
> 
> let's always document the exact error message in tickets

2. Fixed

> I think the key is to NOT set anything in the first step when reference index is generated. From section 3.2.1 :
> 
> > Since 2.4.1a, the annotations can be included on the fly at the mapping step, without including
> > them at the genome generation step.
> 
> "without including" in the index step seems to be the key to not cause a conflict later -- could you try that?

3. Indeed, I have tried and it works. However, as noted below. We will need to unwrap or modify the run_STAR.py wrapper
>1. The star we ran is within an embedded script from gTEX, which did not provide the option to specify the "sjdbOverhang" option.

__________________________________________________________________________________________________________

In conclusion, I think the right way to do is to replace the run_star.py wrapper with our own command. Alternatively, we should modify the run_STAR.py within our docker everytime our docker was built.

**Comment by gaow** [comment]
>  I think the right way to do is to replace the run_star.py wrapper with our own command. Alternatively, we should modify the run_STAR.py within our docker everytime our docker was built.

I agree but could you double check what's special about run_STAR.py that makes it a worthy wrapper to use? We can basically reimplement that if yhou see something useful.

**Comment by hsun3163** [comment]
> > I think the right way to do is to replace the run_star.py wrapper with our own command. Alternatively, we should modify the run_STAR.py within our docker everytime our docker was built.
> 
> I agree but could you double check what's special about run_STAR.py that makes it a worthy wrapper to use? We can basically reimplement that if yhou see something useful.

It is just what GTEX uses, will take a look to see if there is some special treatment.

**Comment by hsun3163** [comment]
Done
